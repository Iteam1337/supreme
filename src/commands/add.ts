import { nvmrc, gitignore, jest, prettierrc } from '../tools'

export type Command =
  | 'gitignore'
  | 'git'
  | 'jest'
  | 'nvm'
  | 'nvmrc'
  | 'prettier'

export const add = (command: Command) => {
  switch (command) {
    case 'nvm':
    case 'nvmrc':
      nvmrc()
      break
    case 'git':
    case 'gitignore':
      gitignore()
      break
    case 'jest':
      jest()
      break
    case 'prettier':
      prettierrc()
      break
    default:
      console.log(`${command} is not a valid command`)
  }
}
