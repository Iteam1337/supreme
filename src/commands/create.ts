export const create = ({ input, flags }) => {
  console.log(input, flags)
}
