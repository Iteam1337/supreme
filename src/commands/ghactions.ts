import { create, createFolder, folderExists, installPkg } from '../utils/file'
import chalk from 'chalk'
import { CLIFlags } from '../'

interface GhActionsProps {
  flags: CLIFlags
}

export const ghactions = async ({ flags }: GhActionsProps) => {
  const githubFolder = folderExists('.github')
  const workflowsFolder = folderExists('.github/workflows')

  installPkg('@semantic-release/changelog')
  installPkg('@semantic-release/git')

  if (!githubFolder) {
    await createFolder('.github')
  }

  if (!workflowsFolder) {
    await createFolder('.github/workflows')
  }

  if (flags.npm) {
    await create({
      templateName: 'ghactions/releaserc',
      output: '.releaserc',
    })
  } else {
    await create({
      templateName: 'ghactions/releaserc.nonpm',
      output: '.releaserc',
    })
  }

  await create({
    templateName: 'ghactions/pr_check.yml',
    output: '.github/workflows/pr_check.yml',
  })

  await create({
    templateName: 'ghactions/release.yml',
    output: '.github/workflows/release.yml',
  })

  console.log(`Added GitHub actions in ${chalk.green(`.github/workflows`)}`)
}
