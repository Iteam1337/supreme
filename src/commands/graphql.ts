import chalk from 'chalk'
import execa from 'execa'
import ora from 'ora'
import path from 'path'
import { husky, jest, nvmrc, prettierrc } from '../tools'
import { create, createFolder, overwrite } from '../utils/file'

interface GraphQLProps {
  flags: any
  name?: string
}

export const graphql = async ({ name, flags }: GraphQLProps) => {
  if (!name) {
    console.log(`No name provided`)
    return
  }

  const projectFolder = { cwd: path.join(process.cwd(), name) }

  const spinner = ora({ text: 'Creating folder', color: 'blue' }).start()

  await createFolder(name)

  spinner.text = 'Creating empty git repository'

  await execa.command('git init', projectFolder)
  await overwrite({
    templateName: 'graphql/gitignore',
    output: `${name}/.gitignore`,
  })

  spinner.text = 'Installing dependencies'

  await create({
    templateName: 'graphql/package.json',
    output: `${name}/package.json`,
    templateData: {
      name,
    },
  })
  await execa.command('npm install --silent', projectFolder)

  await prettierrc({ cwd: name })
  await jest({ cwd: name })
  await nvmrc({ cwd: name })
  await husky({ cwd: name })

  spinner.text = 'Creating base files'

  await create({
    templateName: 'graphql/tsconfig.json',
    output: `${name}/tsconfig.json`,
  })

  await createFolder(`${name}/lib`)

  if (flags.examples) {
    await create({
      templateName: 'graphql/serverExample.ts',
      output: `${name}/lib/server.ts`,
    })

    await createFolder(`${name}/lib/resolvers`)
    await create({
      templateName: 'graphql/resolversExample.ts',
      output: `${name}/lib/resolvers/queue.ts`,
    })

    await createFolder(`${name}/lib/__generated__`)
    await create({
      templateName: 'graphql/graphql.d.ts',
      output: `${name}/lib/__generated__/graphql.d.ts`,
    })
  } else {
    await create({
      templateName: 'graphql/server.ts',
      output: `${name}/lib/server.ts`,
    })
  }

  spinner.stop()

  console.log(chalk.green(`Created app ${name}`))
  console.log(`
  Start the API by running:

  * ${chalk.green(`cd ${name}`)}
  * ${chalk.green('npm run dev')} (starts API on port 4000)
  `)
}
