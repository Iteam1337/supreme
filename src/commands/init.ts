import {
  config,
  gitignore,
  jest,
  nvmrc,
  prettierrc,
  husky,
  eslint,
} from '../tools'
import { CLIFlags } from '../'

interface InitProps {
  flags: CLIFlags
}

export const init = async ({ flags }: InitProps) => {
  await config({ javascript: flags.javascript || false })
  await gitignore()
  await prettierrc()
  await jest()
  await nvmrc()
  await husky()
  await eslint({ node: flags.node, react: flags.react })
}
