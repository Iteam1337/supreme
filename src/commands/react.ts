import execa from 'execa'
import chalk from 'chalk'
import { CLIProps } from '../'

export const react = async ({ name, flags }: CLIProps) => {
  console.log(`Creating app ${chalk.blue(name || '')}`)

  // Create app
  await execa('npx', [
    'create-react-app',
    name || '.',
    flags.typescript ? '--typescript' : '',
  ])

  console.log(chalk.green(`Created app ${name || ''}`))
}
