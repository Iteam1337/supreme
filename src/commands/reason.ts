import execa from 'execa'
import chalk from 'chalk'
import { CLIProps } from '../'
import { create, overwrite, createFolder } from '../utils/file'
import path from 'path'
import ora from 'ora'

export const reason = async ({ name }: CLIProps) => {
  const projectName = name || 'supreme-reason'
  const projectFolder = { cwd: path.join(process.cwd(), projectName) }

  console.log(`Creating app ${chalk.blue(projectName)}`)

  const spinner = ora({ text: 'Creating folder', color: 'blue' }).start()

  // Create app
  await execa.command(`mkdir ${projectName}`)
  await execa.command('mkdir public', projectFolder)
  await execa.command('mkdir src', projectFolder)

  // Overwrite base files
  spinner.text = 'Updating configs'
  await overwrite({
    templateName: 'reason/package.json',
    output: `${projectName}/package.json`,
    templateData: {
      name: projectName,
    },
  })

  await overwrite({
    templateName: 'reason/bsconfig.json',
    output: `${projectName}/bsconfig.json`,
    templateData: {
      name: projectName,
    },
  })

  await overwrite({
    templateName: 'reason/webpack.config.js',
    output: `${projectName}/webpack.config.js`,
  })

  // Add custom README
  await overwrite({
    templateName: 'reason/README.md',
    output: `${projectName}/README.md`,
    templateData: {
      name: projectName,
    },
  })

  // Creating git
  await execa.command('git init', projectFolder)
  await overwrite({
    templateName: 'reason/gitignore',
    output: `${projectName}/.gitignore`,
  })

  // Install dependencies
  spinner.text = 'Installing dependencies'
  await execa.command('npm install --silent', projectFolder)

  // Setup Tailwind CSS
  spinner.text = 'Setting up styling'
  await execa('npx', ['tailwind', 'init'], projectFolder)

  await create({
    templateName: 'reason/postcss.config.js',
    output: `${projectName}/postcss.config.js`,
  })
  await create({
    templateName: 'reason/index.css',
    output: `${projectName}/src/index.css`,
  })
  await create({
    templateName: 'reason/index.js',
    output: `${projectName}/src/index.js`,
  })

  // Creates .releaserc for @semantic-release
  await create({
    templateName: 'reason/releaserc',
    output: `${projectName}/.releaserc`,
  })

  // Move and overwrite index html
  spinner.text = 'Updating base files'

  await overwrite({
    templateName: 'reason/index.html',
    output: `${projectName}/public/index.html`,
    templateData: {
      name: projectName,
    },
  })

  await overwrite({
    templateName: 'reason/Index.re',
    output: `${projectName}/src/Index.re`,
  })

  await create({
    templateName: 'reason/App.re',
    output: `${projectName}/src/App.re`,
  })

  // Test setup
  await createFolder(`${projectName}/__tests__`)

  await create({
    templateName: 'reason/App_test.re',
    output: `${projectName}/__tests__/App_test.re`,
  })

  spinner.stop()

  // Bootstrap done
  console.log(chalk.green(`Created app ${projectName}`))
  console.log(`
	Get your app running by opening two terminal tabs and
	running the following commands:

  * ${chalk.green('npm start')} (starts compiler)
  * ${chalk.green('npm run server')} (start Webpack on port 3000)
`)
}
