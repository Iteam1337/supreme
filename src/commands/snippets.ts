import clipboardy from 'clipboardy'
import { readSnippet } from '../utils/file'

export type SnippetLanguage =
  | 'ts'
  | 'typescript'
  | 'reason'
  | 'reasonml'
  | 'js'
  | 'javascript'

export type SnippetIDE = 'vim' | 'vscode'

interface SnippetsProps {
  language: SnippetLanguage
  ide: SnippetIDE
}

interface CopySnippetProps {
  template: string
  message: string
}

const copySnippet = ({ template, message }: CopySnippetProps) => {
  readSnippet(template).then((snippet) => {
    clipboardy.writeSync(snippet)
    console.log(message)
  })
}

const handleVim = async (language: SnippetLanguage) => {
  switch (language) {
    case 'js':
    case 'javascript':
      copySnippet({
        template: 'javascript.vim',
        message:
          'JavaScript snippets for VIM (UltiSnips) have been copied to the clipboard',
      })
      break

    case 'ts':
    case 'typescript':
      copySnippet({
        template: 'typescript.vim',
        message:
          'Typescript snippets for VIM (UltiSnips) have been copied to the clipboard',
      })
      break

    case 'reason':
    case 'reasonml':
      copySnippet({
        template: 'reason.vim',
        message:
          'ReasonML snippets for VIM (UltiSnips) have been copied to the clipboard',
      })
      break

    default:
      console.log(`I don't have any snippets for that language (${language})`)
  }
}

const handleVscode = async (language: SnippetLanguage) => {
  switch (language) {
    case 'reason':
    case 'reasonml':
      copySnippet({
        template: 'reason.vscode',
        message:
          'ReasonML snippets for VSCode have been copied to the clipboard',
      })
      break

    default:
      console.log(`I don't have any snippets for that language (${language})`)
  }
}

export const snippets = async ({ language, ide }: SnippetsProps) => {
  if (!ide) {
    console.log('--ide flag is missing')
    return
  }

  if (!language) {
    console.log('--language flag is missing')
    return
  }

  switch (ide) {
    case 'vim':
      await handleVim(language)
      break
    case 'vscode':
      await handleVscode(language)
      break
    default:
      console.log(`I don't have any snippets for that IDE (${ide})`)
  }
}
