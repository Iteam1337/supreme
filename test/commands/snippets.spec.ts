import { snippets } from '../../src/commands/snippets'
import clipboardy from 'clipboardy'
import ejs from 'ejs'

jest.mock('ejs')
jest.mock('clipboardy')
jest.spyOn(global.console, 'log').mockImplementation(() => {})

beforeEach(jest.resetAllMocks)

test('should handle undefined ide', async () => {
  await snippets({ language: 'typescript', ide: undefined })

  expect(global.console.log).toHaveBeenCalledWith('--ide flag is missing')
})

test('should handle undefined language', async () => {
  await snippets({ language: undefined, ide: 'vim' })

  expect(global.console.log).toHaveBeenCalledWith('--language flag is missing')
})

test('should handle vim and typescript snippets', async () => {
  ;(ejs as jest.Mock).renderFile.mockResolvedValue('tssnippet')

  await snippets({ language: 'typescript', ide: 'vim' })

  expect((clipboardy.writeSync as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
  expect(global.console.log).toHaveBeenCalledWith(
    'Typescript snippets for VIM (UltiSnips) have been copied to the clipboard'
  )
})

test('should handle vim and typescript (ts) snippets', async () => {
  ;(ejs as jest.Mock).renderFile.mockResolvedValue('tssnippet')

  await snippets({ language: 'ts', ide: 'vim' })

  expect((clipboardy.writeSync as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
  expect(global.console.log).toHaveBeenCalledWith(
    'Typescript snippets for VIM (UltiSnips) have been copied to the clipboard'
  )
})

test('should handle vim and reason snippets', async () => {
  ;(ejs as jest.Mock).renderFile.mockResolvedValue('reasonsnippet')

  await snippets({ language: 'reason', ide: 'vim' })

  expect((clipboardy.writeSync as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
  expect(global.console.log).toHaveBeenCalledWith(
    'ReasonML snippets for VIM (UltiSnips) have been copied to the clipboard'
  )
})

test('should handle vim and reason (reasonml) snippets', async () => {
  ;(ejs as jest.Mock).renderFile.mockResolvedValue('reasonsnippet')

  await snippets({ language: 'reasonml', ide: 'vim' })

  expect((clipboardy.writeSync as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
  expect(global.console.log).toHaveBeenCalledWith(
    'ReasonML snippets for VIM (UltiSnips) have been copied to the clipboard'
  )
})

test('should handle vscode and reason snippets', async () => {
  ;(ejs as jest.Mock).renderFile.mockResolvedValue('reasonsnippet')

  await snippets({ language: 'reason', ide: 'vscode' })

  expect((clipboardy.writeSync as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
  expect(global.console.log).toHaveBeenCalledWith(
    'ReasonML snippets for VSCode have been copied to the clipboard'
  )
})

test('should handle vscode and reason (reasonml) snippets', async () => {
  ;(ejs as jest.Mock).renderFile.mockResolvedValue('reasonsnippet')

  await snippets({ language: 'reasonml', ide: 'vscode' })

  expect((clipboardy.writeSync as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
  expect(global.console.log).toHaveBeenCalledWith(
    'ReasonML snippets for VSCode have been copied to the clipboard'
  )
})

test('should handle vim and javascript snippets', async () => {
  ;(ejs as jest.Mock).renderFile.mockResolvedValue('javascriptsnippet')

  await snippets({ language: 'js', ide: 'vim' })

  expect((clipboardy.writeSync as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
  expect(global.console.log).toHaveBeenCalledWith(
    'JavaScript snippets for VIM (UltiSnips) have been copied to the clipboard'
  )
})

test('should handle vim and javascript (javascript) snippets', async () => {
  ;(ejs as jest.Mock).renderFile.mockResolvedValue('javascriptsnippet')

  await snippets({ language: 'javascript', ide: 'vim' })

  expect((clipboardy.writeSync as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
  expect(global.console.log).toHaveBeenCalledWith(
    'JavaScript snippets for VIM (UltiSnips) have been copied to the clipboard'
  )
})

test('should display error for unknown language', async () => {
  await snippets({ language: '__no_lang__', ide: 'vim' })

  expect(global.console.log).toHaveBeenCalledWith(
    "I don't have any snippets for that language (__no_lang__)"
  )
})

test('should display error for unknown language (vscode)', async () => {
  await snippets({ language: '__no_lang__', ide: 'vscode' })

  expect(global.console.log).toHaveBeenCalledWith(
    "I don't have any snippets for that language (__no_lang__)"
  )
})

test('should display error for unknown IDE', async () => {
  await snippets({ language: 'typescript', ide: '__not_valid__' })

  expect(global.console.log).toHaveBeenCalledWith(
    "I don't have any snippets for that IDE (__not_valid__)"
  )
})
