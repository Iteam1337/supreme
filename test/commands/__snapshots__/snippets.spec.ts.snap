// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle vim and javascript (javascript) snippets 1`] = `"javascriptsnippet"`;

exports[`should handle vim and javascript snippets 1`] = `"javascriptsnippet"`;

exports[`should handle vim and reason (reasonml) snippets 1`] = `"reasonsnippet"`;

exports[`should handle vim and reason snippets 1`] = `"reasonsnippet"`;

exports[`should handle vim and typescript (ts) snippets 1`] = `"tssnippet"`;

exports[`should handle vim and typescript snippets 1`] = `"tssnippet"`;

exports[`should handle vscode and reason (reasonml) snippets 1`] = `"reasonsnippet"`;

exports[`should handle vscode and reason snippets 1`] = `"reasonsnippet"`;
